datasize_macro/src/impl_data_sizes.rs

```diff
@@ -0,0 +1,175 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use proc_macro2::TokenStream as TokenStream2;
+use quote::{format_ident, quote, ToTokens};
+use syn::{Data, DataEnum, DataStruct, DeriveInput, Fields, Ident, Index, Type};
+
+pub fn impl_datasize(input: &DeriveInput) -> TokenStream2 {
+	match &input.data {
+		Data::Enum(e) => impl_datasize_enum(e),
+		Data::Struct(s) => impl_datasize_struct(s),
+		Data::Union(_) => {
+			panic!("Unions are used for C bindings, you probably don't need this trait for it");
+		}
+	}
+}
+
+pub fn impl_static_data_size(input: &DeriveInput) -> TokenStream2 {
+	match &input.data {
+		Data::Enum(e) => impl_static_datasize_enum(e),
+		Data::Struct(s) => impl_static_datasize_struct(s),
+		syn::Data::Union(_) => {
+			panic!("Unions are used for C bindings, you probably don't need this trait for it");
+		}
+	}
+}
+
+fn impl_datasize_enum(data_enum: &DataEnum) -> TokenStream2 {
+	let branches = data_enum
+		.variants
+		.iter()
+		.map(|v| {
+			let ident = &v.ident;
+			match &v.fields {
+				Fields::Named(f) => {
+					// Get a list of all named fields of a named variant
+					let names: Vec<Ident> =
+						f.named.iter().filter_map(|f| f.ident.to_owned()).collect();
```

Comment:
Well i can't do this with only a reference, it give me this error:
```
error[E0507]: cannot move out of `f.ident` which is behind a shared reference
```

And same thing with iterators
```
error[E0382]: use of moved value: `names`
```

Reply:
It shouldn't need to move out of `f.ident` - is that because you're annotating the type explicitly as `Vec<Ident>` (instead of `Vec<_>` or something where it would infer the type and simply annotate for `collect()` that it is a `Vec`)?

Reply:
Nope same issue (changed to `Vec<_>`)
```
error[E0507]: cannot move out of `f.ident` which is behind a shared reference
  --> datasize_macro/src/impl_data_sizes.rs:41:23
   |
41 |                         .filter_map(|f| f.ident)
   |                                         ^^^^^^^ move occurs because `f.ident` has type `Option<proc_macro2::Ident>`, which does not implement the `Copy` trait
```

Reply:
Ohh yeah my mistake,
```rust
fields.named.iter().filter_map(|field| &field.ident)
```
forgot to actually, yknow, do the reference bit :P